accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/helper/QueryHelper.java

```diff
@@ -30,202 +32,213 @@
+            RepositoryException {
+        Set<Node> nodes = new HashSet<Node>();
+
+        Query query = session.getWorkspace().getQueryManager()
+                .createQuery(queryStatement, queryLanguageType);
+        QueryResult queryResult = query.execute();
+        NodeIterator nit = queryResult.getNodes();
+        List<String> paths = new ArrayList<String>();
+
+        while (nit.hasNext()) {
+            // get the next rep:policy node
+            Node node = nit.nextNode();
+            // AcHelper.LOG.debug("adding node: {} to node set", node.getPath());
+            paths.add(node.getPath());
+            nodes.add(node);
+        }
+        return nodes;
+    }
+
+    public static Set<String> getUsersFromHome(final Session session)
+            throws InvalidQueryException, RepositoryException {
+        Set<String> users = getPrincipalsFromHome(session, NT_REP_USER);
+        return users;
+    }
+
+    public static Set<String> getGroupsFromHome(final Session session)
+            throws InvalidQueryException, RepositoryException {
+        Set<String> groups = getPrincipalsFromHome(session, NT_REP_GROUP);
+        return groups;
+    }
+
+    private static Set<String> getPrincipalsFromHome(final Session session, String principalNodeType)
+            throws InvalidQueryException, RepositoryException {
+        Set<String> principals = new TreeSet<String>();
+        String queryStringPrincipals = "SELECT * FROM [" + principalNodeType + "]";
+        Query queryPrincipals = session.getWorkspace().getQueryManager()
+                .createQuery(queryStringPrincipals, Query.JCR_SQL2);
+        QueryResult queryResultPrincipals = queryPrincipals.execute();
+        NodeIterator nitPrincipals = queryResultPrincipals.getNodes();
+
+        while (nitPrincipals.hasNext()) {
+            Node node = nitPrincipals.nextNode();
+            String tmp = node.getProperty("rep:principalName").getString();
+            principals.add(tmp);
+        }
+        return principals;
+    }
+
+    public static Set<AclBean> getAuthorizablesAcls(final Session session,
+            final Set<String> authorizableIds, Set<String> principalIdsToBeFilled) throws InvalidQueryException,
+            RepositoryException {
+        
+        LOG.debug("Querying AclBeans for {} authorizables", authorizableIds.size());
+
+        StopWatch sw = new StopWatch();
+        sw.start();
+        
+        Set<Node> nodeSet = new LinkedHashSet<Node>();
+
+        Iterator<String> authorizablesIdIterator = authorizableIds.iterator();
+
+
+        while (authorizablesIdIterator.hasNext()) {
+            StringBuilder queryStringBuilder = new StringBuilder();
+            queryStringBuilder.append(
+                    "SELECT ace.* FROM [rep:ACE] AS ace INNER JOIN [rep:Authorizable] AS authorizable "
+                            + "ON ace.[rep:principalName] = authorizable.[rep:principalName] WHERE ");
```

Comment:
So here we have to do a join. Performance of it is better than I expected (did some test in the groovy console that made the difference almost neglectable).